### ISSUE TITLE
Faster performance measures from package Modelmetrics

### ISSUE THREAD

**Issue by studerus** [mlr-org/mlr#1319]
I noticed that package `caret` uses in its newest version the auc measure from package ModelMetrics, because it's programmed with Rcpp. ModelMetrics has many more performance measures all implemented in C++.

I did a quick speed comparison between our implementation of auc and the one in ModelMetrics. The one in ModelMetrics seems to be more than 10 times as fast.

```
library(mlr)
library(ModelMetrics)
library(microbenchmark)

pred <- predict(train(makeLearner('classif.logreg', predict.type = 'prob'), pid.task), pid.task)
prob <- getPredictionProbabilities(pred)
truth <- getPredictionTruth(pred)

microbenchmark(
  ModelMetrics = ModelMetrics::auc(truth, prob),
  mlr = measureAUC(prob, truth, negative = 'neg', positive = 'pos'))
```

```
Unit: microseconds
         expr      min       lq     mean   median       uq      max neval cld
 ModelMetrics  206.438  216.555  233.762  223.446  238.987  376.809   100  a 
          mlr 2657.890 2737.211 3026.499 2899.223 3066.808 5915.154   100   b
```

**Comment by catastrophic-failure** [comment]
`measureAUC` is based on `ROCR`, the rest of the AUC metrics are based on `caTools`'s `colAUC` which I suppose must be faster.

**Comment by larskotthoff** [comment]
Good catch! Could you make a PR for the `ModelMetrics` measures please?

**Comment by PhilippPro** [comment]
`colAUC` is only twice as slow as the `ModelMetrics` package.

```
library(mlr)
library(ModelMetrics)
library(microbenchmark)
library(caTools)

pred <- predict(train(makeLearner('classif.logreg', predict.type = 'prob'), pid.task), pid.task)
prob <- getPredictionProbabilities(pred)
truth <- getPredictionTruth(pred)

microbenchmark(
  ModelMetrics = ModelMetrics::auc(truth, prob),
  mlr = measureAUC(prob, truth, negative = 'neg', positive = 'pos'),
  mlr2 = colAUC(prob, truth))
```
```

Unit: microseconds
         expr      min        lq      mean    median       uq      max neval cld
 ModelMetrics  320.223  338.3465  375.2747  355.4225  413.216  485.817   100 a  
          mlr 3501.354 3589.9820 4306.6350 3743.3890 5121.008 7983.279   100   c
         mlr2  818.959  861.1435  945.5868  897.6010 1029.182 1510.948   100  b 

```

**Comment by studerus** [comment]
Sorry, I'm too busy at the moment to implement it myself. I just wanted to make a suggestion for improvement.

**Comment by PhilippPro** [comment]
I made a somewhat different comparison with bigger vectors (it is a scenario with 7 million observations) and here colAUC is nearly twice as fast as the ModelMetrics and our original implementation. I will make a PR, so we use colAUC instead of the ROCR implementation. (I slightly have to change colAUC, as it takes pmax(auc, 1 - auc) and I think this is not desirable for the twoclass case).

Simulation:

```
> 
> library(mlr)
> library(ModelMetrics)
> library(caTools)
> 
> pred = predict(train(makeLearner('classif.logreg', predict.type = 'prob'), pid.task), pid.task)
> prob = getPredictionProbabilities(pred)
> truth = getPredictionTruth(pred)
> 
> prob = rep(prob, 10000)
> truth = rep(truth, 10000)
> system.time(ModelMetrics <- ModelMetrics::auc(truth, prob))
       User      System verstrichen 
     24.808       0.000      24.825 
> system.time(mlr2 <- mlr:::colAUC(prob, truth))
       User      System verstrichen 
     14.148       0.060      14.221 
> system.time(mlr <- measureAUC(prob, truth, negative = 'neg', positive = 'pos'))
       User      System verstrichen 
     26.032       0.108      26.153 
```

**Comment by JackStat** [comment]
I opened a pull request to use ModelMetrics as you can see in the reference above.  Didn't even know about this :) 

@PhilippPro I am unable to replicate your results for some reason. Here were my results with your code

    > library(mlr)
    > library(ModelMetrics)
    > library(caTools)
    > 
    > pred = predict(train(makeLearner('classif.logreg', predict.type = 'prob'), pid.task), pid.task)
    > prob = getPredictionProbabilities(pred)
    > truth = getPredictionTruth(pred)
    > 
    > prob = rep(prob, 10000)
    > truth = rep(truth, 10000)
    > system.time(ModelMetrics <- auc(truth, prob))
       user  system elapsed 
      3.135   0.087   3.231 
    > 
    > system.time(mlr2 <- colAUC(prob, truth))
       user  system elapsed 
      6.722   0.398   7.154 
    >      
    > system.time(mlr <- measureAUC(prob, truth, negative = 'neg', positive = 'pos'))
       user  system elapsed 
      6.988   0.288   7.297

**Comment by PhilippPro** [comment]
Ok, very strange, what system and what package version of ModelMetrics are you using? I launched it on another pc and got around 9 sec for each method, I am using Debian on both systems and the developer version of ModelMetrics. 
The "system" time is always faster for ModelMetrics, also in my experiments.

**Comment by JackStat** [comment]
I am on a Macbook Pro 16GB RAM and I am using the developer version of ModelMetrics

**Comment by JackStat** [comment]
I just added omp to the auc.cpp functions and it sped it up a bit more in ModelMetrics. Try reinstalling and running the test again.

**Comment by larskotthoff** [comment]
Could you run on a range of different inputs (~10) please so we get a better idea of what's going on?

**Comment by PhilippPro** [comment]
Ok, I changed a bit the experimental setup (no bindings, thx lars) and increased the number of observations (30 Million observations are not very common, though, and if so you will not make 10 reps of 10 CV in mlr ;)). 
Maybe a bit more important in a multiclassAUC case, where many combinations of AUCs have to be computed. 

```
> pred = predict(train(makeLearner('classif.logreg', predict.type = 'prob'), pid.task), pid.task)
> prob = getPredictionProbabilities(pred)
> truth = getPredictionTruth(pred)
> 
> prob = runif(4 * 7680000, 0, 1)
> truth = rep(truth, 40000)
> system.time(ModelMetrics <- auc(truth, prob))
       User      System verstrichen 
      20.63        0.34       21.00 
> system.time(mlr2 <- mlr:::colAUC(prob, truth))
       User      System verstrichen 
      29.75        1.07       30.84 
> system.time(mlr <- measureAUC(prob, truth, negative = 'neg', positive = 'pos'))
       User      System verstrichen 
      35.09        1.41       36.54 
```

**Comment by larskotthoff** [comment]
Let's all run this code please:

```
library(mlr)
library(ModelMetrics)
library(microbenchmark)

pred = predict(train(makeLearner('classif.logreg', predict.type = 'prob'), pid.task), pid.task)
prob = getPredictionProbabilities(pred)
truth = getPredictionTruth(pred)

for (exp in 1:3) {
    print(paste("Length", length(prob) * 10^exp))

    prob.rep = rep(prob, 10^exp)
    truth.rep = rep(truth, 10^exp)

    print(microbenchmark(
        ModelMetrics = auc(truth.rep, prob.rep),
        mlr.colAUC = mlr:::colAUC(prob.rep, truth.rep),
        mlr.measureAUC = measureAUC(prob.rep, truth.rep, negative = 'neg', positive = 'pos')))
}
```

This is what I'm getting with the latest dev version of ModelMetrics:

```
[1] "Length 7680"
Unit: milliseconds
           expr      min       lq     mean   median       uq       max neval
   ModelMetrics 3.350211 3.431984 3.614032 3.479467 3.663359  5.082249   100
     mlr.colAUC 2.440899 2.575904 4.053664 2.804450 3.900653 84.118785   100
 mlr.measureAUC 4.320556 4.604280 5.777352 5.781648 6.149680 34.220524   100
 cld
  a 
  a 
   b
[1] "Length 76800"
Unit: milliseconds
           expr      min       lq     mean   median        uq      max neval
   ModelMetrics 37.54787 39.35613 52.76276 40.66295  43.40542 130.0693   100
     mlr.colAUC 28.81107 31.63644 61.11806 33.74734 114.25657 197.7511   100
 mlr.measureAUC 31.19532 34.62176 66.59005 37.05100 118.64481 151.4802   100
 cld
  a 
  ab
   b
[1] "Length 768000"
Unit: milliseconds
           expr      min       lq     mean   median       uq      max neval cld
   ModelMetrics 472.1781 524.3678 550.0549 528.5596 579.4231 716.6749   100  a 
     mlr.colAUC 433.1590 529.6005 551.0527 541.6656 568.3904 666.7950   100  a 
 mlr.measureAUC 454.3492 679.8277 693.4670 689.8580 724.5738 976.9709   100   b
```

Looks like ModelMetrics and colAUC are about equal to me.

**Comment by JackStat** [comment]
[1] "Length 7680"
    Unit: milliseconds
               expr      min       lq     mean   median       uq       max neval
       ModelMetrics 1.763520 1.935771 2.100617 2.007341 2.181299  3.567405   100
         mlr.colAUC 1.971687 2.389428 2.644324 2.566596 2.830913  5.914579   100
     mlr.measureAUC 3.042056 3.606622 4.233706 3.762715 4.092706 30.055935   100
    [1] "Length 76800"
    Unit: milliseconds
               expr      min       lq     mean   median       uq      max neval
       ModelMetrics 16.70773 17.03178 18.41149 17.28649 19.13297 32.06292   100
         mlr.colAUC 23.71442 24.27733 26.84801 24.95892 28.32660 42.93649   100
     mlr.measureAUC 34.01665 35.37086 38.04786 36.37308 39.58566 55.06542   100
    [1] "Length 768000"
    Unit: milliseconds
               expr      min       lq     mean   median       uq      max neval
       ModelMetrics 187.1793 210.8090 225.8559 221.4361 235.5894 321.8677   100
         mlr.colAUC 311.4347 342.9822 367.1453 357.9477 379.3303 560.0892   100
     mlr.measureAUC 436.8680 465.3596 495.4876 484.9330 512.7303 654.6803   100

**Comment by PhilippPro** [comment]
On my windows machine: 
```
[1] "Length 7680"
Unit: milliseconds
           expr      min       lq     mean   median       uq      max neval
   ModelMetrics 1.395380 1.419718 1.452124 1.424526 1.452469 2.488488   100
     mlr.colAUC 1.515869 1.534648 1.707634 1.552075 1.626141 4.164506   100
 mlr.measureAUC 2.931681 2.951662 3.114895 2.965484 3.070196 4.209877   100
[1] "Length 76800"
Unit: milliseconds
           expr      min       lq     mean   median       uq       max neval
   ModelMetrics 15.79147 16.30783 18.46116 16.71121 17.83872  87.57391   100
     mlr.colAUC 19.22523 19.93269 24.14198 20.34824 21.39027 100.04760   100
 mlr.measureAUC 34.16187 35.09858 38.70682 36.04626 36.95939 106.82498   100
[1] "Length 768000"
Unit: milliseconds
           expr      min       lq     mean   median       uq      max neval
   ModelMetrics 196.4887 203.4115 225.9707 207.6493 257.4479 391.3022   100
     mlr.colAUC 243.8877 259.2699 302.2672 313.6780 324.4290 434.1155   100
 mlr.measureAUC 395.6229 451.4085 468.7522 470.4891 483.4330 681.6079   100
```

On my Debian machine:

```
[1] "Length 7680"
Unit: milliseconds
           expr      min       lq     mean   median       uq        max neval cld
   ModelMetrics 3.908389 3.944078  4.43913 3.983085 4.712507   8.180790   100  a 
     mlr.colAUC 3.233792 3.322805  3.92493 3.827478 3.974703   9.140898   100  a 
 mlr.measureAUC 6.156300 6.324478 12.25661 7.302186 8.549971 366.653540   100   b
[1] "Length 76800"
Unit: milliseconds
           expr      min       lq     mean   median       uq      max neval cld
   ModelMetrics 42.80586 43.10932 45.49860 44.16410 44.63602 166.1305   100  a 
     mlr.colAUC 33.91458 35.66739 52.96697 36.63196 38.13662 175.5014   100  ab
 mlr.measureAUC 41.44158 43.22323 57.21239 44.06475 45.64254 167.8408   100   b
[1] "Length 768000"
Unit: milliseconds
           expr      min       lq     mean   median       uq       max neval cld
   ModelMetrics 487.3340 490.7104 518.0633 493.5199 495.5739  786.0708   100 a  
     mlr.colAUC 434.7209 628.5557 626.7171 635.5901 644.2319 1259.7223   100  b 
 mlr.measureAUC 569.3901 758.9822 756.1882 764.0426 771.2568 1165.7695   100   c
```

**Comment by studerus** [comment]
It seems that the performance gain of Modelmetrics is minimal when the sample is very large . However, when the sample is of the size of the pid.task, which is probably more commonly the case, Modelmetrics is still about 3 times as fast on my computer. I had to increase the evaluation times to 10000 to get more stable results:

```
microbenchmark(
  ModelMetrics = ModelMetrics::auc(truth, prob),
  mlr = measureAUC(prob, truth, negative = 'neg', positive = 'pos'),
  times = 10000)
```

```
Unit: microseconds
         expr     min       lq     mean  median      uq      max neval cld
 ModelMetrics 183.539 199.9915  328.040 209.198 234.858 84573.80 10000  a 
          mlr 617.028 652.6490 1098.322 689.779 769.171 88063.15 10000   b
```

On the other hand, one could argue that we should not care that much about the performance in small and medium sized samples, as the user would probably not notice the difference, except if he performs highly complex nested-cross validation procedures with several tens of thousands model evaluations.

I also see the point that we should not introduce unnecessary additional dependencies. However, there are additional performance that could be replaced with a ModelMetrics version. For example, the brier score is another candidate, although the absolute time here is relatively small compared to the auc:

```
microbenchmark(
  ModelMetrics = ModelMetrics::brier(truth, prob),
  mlr = measureBrier(prob, truth, negative = 'neg', positive = 'pos'),
  times = 10000)
```

```
Unit: microseconds
         expr    min     lq     mean median     uq        max neval cld
 ModelMetrics  9.057 10.566 24.60009 12.075 12.981 105662.615 10000   a
          mlr 28.679 30.792 38.80970 32.000 35.320   5159.002 10000   a
```

**Comment by JackStat** [comment]
You should drop the `ModelMetrics::brier` and instead just use `brier`.  There is a little overhead when you do that and it takes a bit of time.

**Comment by studerus** [comment]
Sorry, I realized that my comparison above was unfair, because I did not use the mlr:::AUC version. With this, the Modelmetrix version is "only" 2 times as fast.

**Comment by studerus** [comment]
Thanks, JackStat, for pointing that out.

**Comment by studerus** [comment]
I created my own auc function by stealing some code from the `MKmisc::auc` function. It's purely R based and quite a bit faster than ModelMetrics with large vectors.

```
library(mlr)
library(ModelMetrics)
library(microbenchmark)

pred = predict(train(makeLearner('classif.logreg', predict.type = 'prob'), pid.task), pid.task)
prob = getPredictionProbabilities(pred)
truth = getPredictionTruth(pred)

myAUC <- function(probabilities, truth, negative, positive) {
  y <- sort(probabilities[truth == positive])
  x <- sort(probabilities[truth == negative])
  nx <- length(x)
  ny <- length(y)
  (nx * ny + nx * (nx + 1)/2 - sum(rank(c(x, y))[1:nx]))/(nx * ny)
} 

all.equal(myAUC(prob, truth, negative = 'neg', positive = 'pos'),
    measureAUC(prob, truth, negative = 'neg', positive = 'pos'))

for (exp in 1:3) {
    print(paste("Length", length(prob) * 10^exp))

    prob.rep = rep(prob, 10^exp)
    truth.rep = rep(truth, 10^exp)

    print(microbenchmark(
        ModelMetrics = auc(truth.rep, prob.rep),
        myAUC = myAUC(prob.rep, truth.rep, negative = 'neg', positive = 'pos')))
}
```
```
[1] "Length 7680"
Unit: milliseconds
         expr      min       lq     mean   median       uq       max neval cld
 ModelMetrics 1.460762 1.557815 1.799155 1.638111 1.804142  3.900797   100   a
        myAUC 1.353899 1.458649 3.398711 1.538645 1.699241 82.474885   100   a
[1] "Length 76800"
Unit: milliseconds
         expr      min       lq     mean   median       uq      max neval cld
 ModelMetrics 21.37167 23.79721 42.33579 24.77015 26.67617 187.3647   100   b
        myAUC 14.58195 16.77641 29.88384 18.02617 19.48240 186.8431   100  a 
[1] "Length 768000"
Unit: milliseconds
         expr      min       lq     mean   median       uq       max neval cld
 ModelMetrics 566.4519 687.8946 765.5887 761.6459 831.1141 1078.8367   100   b
        myAUC 169.2201 267.5500 332.1705 337.6118 347.2933  583.2907   100  a 
```

**Comment by JackStat** [comment]
The `myAUC` function is generating this warning `In nx * ny : NAs produced by integer overflow`.  `Metrics::auc` and `MLmetrics::AUC` have that same issue and that was the original reason that I built ModelMetrics. I think you can overcome the error by rewriting that function a bit

    myAUC <- function(probabilities, truth, negative, positive) {
      y <- sort(probabilities[truth == positive])
      x <- sort(probabilities[truth == negative])
      nx <- as.numeric(length(x))
      ny <- as.numeric(length(y))
      (nx * ny + nx * (nx + 1)/2 - sum(rank(c(x, y))[1:nx]))/(nx * ny)
    }

**Comment by studerus** [comment]
Thanks. Yes, I noticed the same thing and came up with the same solution. :-)

**Comment by larskotthoff** [comment]
Ok, so it seems to me that at least on Linux there's no significant performance gain to using ModelMetrics and the times are so small that we probably don't care anyway.

Thanks everybody for investigating, shall we close here?

**Comment by studerus** [comment]
I improved my function a little bit. It's still purely R based without any package dependencies and it's consistently about 3-4 times faster then our current implementation. I checked four different sample sizes and two platforms (windows and linux). I therefore suggest to replace our current implementation with my version. I also realized that I have to use measureAUC for a fair comparison, not mlr:::colAUC, because the latter does not check which factor level is the positive class.

```
library(mlr)
library(microbenchmark)

pred = predict(train(makeLearner('classif.logreg', predict.type = 'prob'), pid.task), pid.task)
prob = getPredictionProbabilities(pred)
truth = getPredictionTruth(pred)

myAUC <- function(probabilities, truth, negative, positive) {
  pos.level = which(levels(truth) == positive)
  i = as.integer(truth) == pos.level
  y = probabilities[i]
  x = probabilities[!i]
  y = sort.default(y)
  x = sort.default(x)
  nx = as.numeric(length(x))
  ny = as.numeric(length(y))
  (nx * ny + nx * (nx + 1)/2 - sum(rank(c(x, y))[1:nx]))/(nx * ny)
}

for (i in 0:3) {
	prob.rep = rep(prob, 10^i)
	truth.rep = rep(truth, 10^i)
	cat('\n', 10^i, 'times the original size\n')
	print(microbenchmark(
        myAUC = myAUC(prob.rep, truth.rep, negative = 'neg', positive = 'pos'),
        mlr = measureAUC(prob.rep, truth.rep, negative = 'neg', positive = 'pos'),
        times = 10^(4 - i)))
}
```

On windows 10
```
 1 times the original size
Unit: microseconds
  expr     min       lq     mean  median      uq        max neval cld
 myAUC 120.814 131.9565 144.0572 135.476 139.581   4605.867 10000  a 
   mlr 678.843 700.5420 751.5302 713.445 732.798 112287.902 10000   b

 10 times the original size
Unit: milliseconds
  expr      min       lq     mean   median       uq      max neval cld
 myAUC 1.130427 1.169868 1.244031 1.187315 1.220451 4.449278  1000  a 
   mlr 3.931716 4.057662 4.236621 4.144605 4.245919 7.520932  1000   b

 100 times the original size
Unit: milliseconds
  expr      min       lq     mean   median       uq       max neval cld
 myAUC 11.54443 11.88869 12.84104 12.17284 13.55486  17.93933   100  a 
   mlr 42.39849 42.99830 45.74776 44.54733 45.81411 151.42443   100   b

 1000 times the original size
Unit: milliseconds
  expr      min       lq     mean   median       uq     max neval cld
 myAUC 134.0068 141.0318 199.4332 198.8922 252.4869 284.835    10  a 
   mlr 568.1793 574.9568 651.3081 677.0273 685.9032 705.109    10   b
```

On Linux Ubuntu
```
 1 times the original size
Unit: microseconds
  expr     min       lq     mean  median      uq        max neval cld
 myAUC 120.814 131.9565 144.0572 135.476 139.581   4605.867 10000  a 
   mlr 678.843 700.5420 751.5302 713.445 732.798 112287.902 10000   b

 10 times the original size
Unit: milliseconds
  expr      min       lq     mean   median       uq      max neval cld
 myAUC 1.130427 1.169868 1.244031 1.187315 1.220451 4.449278  1000  a 
   mlr 3.931716 4.057662 4.236621 4.144605 4.245919 7.520932  1000   b

 100 times the original size
Unit: milliseconds
  expr      min       lq     mean   median       uq       max neval cld
 myAUC 11.54443 11.88869 12.84104 12.17284 13.55486  17.93933   100  a 
   mlr 42.39849 42.99830 45.74776 44.54733 45.81411 151.42443   100   b

 1000 times the original size
Unit: milliseconds
  expr      min       lq     mean   median       uq     max neval cld
 myAUC 134.0068 141.0318 199.4332 198.8922 252.4869 284.835    10  a 
   mlr 568.1793 574.9568 651.3081 677.0273 685.9032 705.109    10   b
```

**Comment by larskotthoff** [comment]
Thanks, could you make a pull request for that please?

**Comment by studerus** [comment]
An even more efficient calculation would be possible using fast ranking function of data.table, although it's slightly slower with small and medium sized data sets. However, I think my first proposal is good enough. I post this just for the record.

```
library(mlr)
library(data.table)
library(microbenchmark)

pred = predict(train(makeLearner('classif.logreg', predict.type = 'prob'), pid.task), pid.task)
prob = getPredictionProbabilities(pred)
truth = getPredictionTruth(pred)

myAUC <- function(probabilities, truth, negative, positive) {
  if (is.logical(truth)) {
     i = truth == positive
  } else {
     pos.level = which(levels(truth) == positive)
     i = as.integer(truth) == pos.level
  }
  y = probabilities[i]
  x = probabilities[!i]
  y = sort.int(y)
  x = sort.int(x)
  nx = as.numeric(length(x))
  ny = as.numeric(length(y))
  (nx * ny + nx * (nx + 1)/2 - sum(rank(c(x, y))[1:nx]))/(nx * ny)
}

myAUC2 <- function(probabilities, truth, negative, positive) {
  if (is.logical(truth)) {
     i = truth == positive
  } else {
     pos.level = which(levels(truth) == positive)
     i = as.integer(truth) == pos.level
  }
  r = frankv(r)
  n_pos = as.numeric(sum(i))
  n_neg = as.numeric(sum(!i))
  (sum(r[i]) - n_pos * (n_pos + 1)/2)/(n_pos * n_neg)
}

for (i in 0:3) {
	prob.rep = rep(prob, 10^i)
	truth.rep = rep(truth, 10^i)
	cat('\n', 10^i, 'times the original size\n')
	print(microbenchmark(
        myAUC = myAUC(prob.rep, truth.rep, negative = 'neg', positive = 'pos'),
        myAUC2 = myAUC2(prob.rep, truth.rep, negative = 'neg', positive = 'pos'),
        times = 10^(4 - i)))
}

 1 times the original size
Unit: microseconds
   expr     min      lq     mean  median      uq       max neval cld
  myAUC 109.581 122.561 215.2558 128.599 147.918  80407.96 10000  a 
 myAUC2 195.614 225.500 327.9919 239.386 263.837 103596.30 10000   b

 10 times the original size
Unit: microseconds
   expr      min       lq     mean    median       uq       max neval cld
  myAUC 1049.007 1100.477 1764.902 1138.8145 1230.735 83826.065  1000   b
 myAUC2  681.931  736.268  822.188  777.6245  846.451  3301.278  1000  a 

 100 times the original size
Unit: milliseconds
   expr       min        lq      mean    median        uq       max neval cld
  myAUC 10.678136 11.429648 15.989894 11.843213 13.188508  90.77034   100   b
 myAUC2  4.104863  4.492316  5.826988  4.695929  4.849583 100.80005   100  a 

 1000 times the original size
Unit: milliseconds
   expr       min        lq      mean    median        uq       max neval cld
  myAUC 134.70547 226.87528 273.07077 300.67239 311.38781 387.34587    10   b
 myAUC2  69.79624  71.42454  73.06875  72.87851  74.75842  77.13778    10  a 
```

**Comment by JackStat** [comment]
Interesting.  I think `frankv` is the key here. You should use this function

**Comment by JackStat** [comment]
I went back and wrote the `auc` function in `ModelMetrics` using the  `frankv` function from `data.table`.  Thanks for the great deep dive on this guys.

**Comment by studerus** [comment]
Yes, the ranking is by far the most time consuming factor as I found out through `profvis`. However, the `frankv` function seems to be only faster when the vector is bigger than about 10000. In my newest auc version (f304bab6130bc66d289020abc7c296203647ed11), I therefore use frankv conditional on the length of the vector.

**Comment by pat-s** [comment]
Solved by #1359.
